### PR TITLE
Elasticache core module

### PR DESCRIPTION
**Pull request by ummiyah-moe**

### CODE REVIEW

File: elasticache/locals.tf [hunk 1]
```diff
@@ -0,0 +1,16 @@
+locals {
+  _subservice_name     = coalesce(var.subservice_name, var.service_name)
```
Comment (sudip-moengage) [inline elasticache/locals.tf:2]: subservice_name is not optional, so you can remove this line and just use `var.subservice_name` everywhere.

File: elasticache/locals.tf [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  _resource_identifier = "${var.business_name}-${var.service_name}-${local._subservice_name}"
+  resource_identifier  = lower(local._resource_identifier)
+
+  default_tags = {
+    ManagedBy   = "terraform"
+    CreatedBy   = lower(var.created_by)
+    Environment = lower(var.environment)
+    Business    = lower(var.business_name)
+    Service     = lower(var.service_name)
+    SubService  = lower(local._subservice_name)
```
Comment (sudip-moengage) [inline elasticache/locals.tf:12]: var.subservice_name

File: elasticache/variables.tf [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+variable "auto_minor_version_upgrade" {
+  description = "Specifies whether a minor engine upgrades will be applied automatically to the underlying Cache Cluster instances during the maintenance window"
+  type        = bool
+  default     = true
+}
+
+variable "availability_zones" {
+  description = "A list of EC2 availability zones in which the replication group's cache clusters will be created. The order of the availability zones in the list is not important"
+  type        = list(string)
+  default     = []
```
Comment (sudip-moengage) [inline elasticache/variables.tf:83]: It should be mandatory I guess?

File: elasticache/variables.tf [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+variable "at_rest_encryption_enabled" {
+  description = "Whether to enable encryption at rest"
+  type        = bool
+  default     = false
```
Comment (sudip-moengage) [inline elasticache/variables.tf:89]: default true

File: elasticache/variables.tf [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+variable "kms_key_id" {
+  description = "The ARN of the key that you wish to use if encrypting at rest. If not supplied, uses service managed encryption. Can be specified only if at_rest_encryption_enabled = true"
+  type        = string
+  default     = ""
```
Comment (sudip-moengage) [inline elasticache/variables.tf:95]: enable this, if needed create kms key. Security team wants us to encrypt everything we have at rest.

File: elasticache/variables.tf [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+variable "subservice_name" {
+  type        = string
+  description = "Sub Service name"
+}
+
+variable "created_by" {
+  type        = string
+  description = "Created by"
```
Comment (sudip-moengage) [inline elasticache/variables.tf:150]: let's deprecate this field. make it default=""

**Review by sudip-moengage — CHANGES_REQUESTED** [review]
Few other changes:
- Move this module to `elasticache/redis` directory
- Please add an example in terragrunt repo: https://github.com/moengage/terragrunt/tree/master/staging/ap-south-1/default/sre/examples

File: elasticache/locals.tf (hunk 1 above)
Comment (ummiyah-moe) [reply elasticache/locals.tf]: Updated.

File: elasticache/variables.tf (hunk 6 above)
Comment (ummiyah-moe) [reply elasticache/variables.tf]: removed this field from module.

File: elasticache/variables.tf (hunk 5 above)
Comment (ummiyah-moe) [reply elasticache/variables.tf]: used aws default kms key.

**Comment by ummiyah-moe** [comment]
@sudip-moengage  1. Fixed dir and comments
2. https://github.com/moengage/terragrunt/pull/103 will update source of terragrunt after this PR is merged.

**Pull request merged**